dali/operators/generic/transpose/transpose.h

```diff
@@ -74,18 +79,27 @@ class Transpose : public Operator<Backend> {
   bool SetupImpl(std::vector<OutputDesc> &output_desc,
                  const Workspace &ws) override {
     const auto &input = ws.Input<Backend>(0);
-    SetOutputLayout(input);
-
     const auto &input_shape = input.shape();
-    int dim = input_shape.sample_dim();
-    int pdim = perm_.size();
-    DALI_ENFORCE(dim == pdim, make_string("Input has different dimensionality (", dim,
-        ") than the length of the permutation (", pdim, ")"));
+    int ndim = input_shape.sample_dim();
+
+    if (default_perm_ && static_cast<int>(perm_.size()) != ndim) {
+      perm_.resize(ndim);
+      std::iota(perm_.begin(), perm_.end(), 0);
+      std::reverse(perm_.begin(), perm_.end());
+    }
 
+    SetOutputLayout(input);
     output_desc.resize(1);
-    permute_dims(output_desc[0].shape, input_shape, make_span(perm_));
     output_desc[0].type = input.type();
-
+    output_desc[0].shape = input_shape;
+
+    if (!perm_.empty()) {
+      int dim = input_shape.sample_dim();
+      int pdim = perm_.size();
+      DALI_ENFORCE(dim == pdim, make_string("Input has different dimensionality (", dim,
+          ") than the length of the permutation (", pdim, ")"));
+      permute_dims(output_desc[0].shape, input_shape, perm_);
```

Comment:
What happened before? Some error later in the kernel?

Reply:
No, this was just moved. I moved to after I potentially regenerate the perm values.